### PR TITLE
Stabilize pub(restricted)

### PR DESCRIPTION
**Pull request by cramertj**
Fix https://github.com/rust-lang/rust/issues/32409

### CODE REVIEW

**Comment by cramertj** [comment]
@retep998 That seems reasonable to me.

**Comment by cramertj** [comment]
@retep998 I'll have a new commit up in a minute with that change.

File: src/librustc_privacy/lib.rs [hunk 1]
```diff
@@ -951,18 +949,10 @@ impl<'a, 'tcx: 'a> TypeVisitor<'tcx> for SearchInterfaceForPrivateItemsVisitor<'
...
-                        err.span_label(self.span, &format!("can't leak private type"));
-                        err.emit();
-                    } else {
-                        self.tcx.sess.add_lint(lint::builtin::PRIVATE_IN_PUBLIC,
-                                               node_id,
-                                               self.span,
-                                               format!("private type `{}` in public \
-                                                        interface (error E0446)", ty));
-                    }
+                    let mut err = struct_span_err!(self.tcx.sess, self.span, E0446,
```
Comment (petrochenkov) [inline src/librustc_privacy/lib.rs]: `private_in_public` can't be turned into an error in the near future unfortunately.
`self.tcx.sess.features.borrow().pub_restricted` can be well approximated with `is_pub_restricted(vis) && is_pub_restricted(self.required_visibility)` where `is_pub_restricted = not pub && not private`.
The idea is to always report an error and not warning when `pub(something)` is involved.

Comment (cramertj) [inline src/librustc_privacy/lib.rs]: What do you mean by `not private`? `Visibility` is either `Public`, `Restricted(DefId)`, or `Invisible` (reserved for private external items).

Comment (petrochenkov) [inline src/librustc_privacy/lib.rs]: You can use `hir::Visibility` for this (as opposed to `ty::Visibility`).
For `vis` HIR visibility is immediately available (`item.vis`) , for `self.required_visibility` passed from `PrivateItemsInPublicInterfacesVisitor::visit_item` to here somehow.

Comment (cramertj) [inline src/librustc_privacy/lib.rs]: Even for `hir::Visibility`, I still only see `Public`, `Crate`, `Restricted {...}`, and `Inherited`, no `Private`. Am I missing something obvious?

**Comment by sfackler** [comment]
Beta nominating since I think @aturon was interested in landing this in 1.17.

File: src/librustc_privacy/lib.rs (hunk 1 above)
Comment (petrochenkov) [inline src/librustc_privacy/lib.rs]: `Inherited` is private (i.e. no any explicit `pub`)

**Comment by nikomatsakis** [comment]
I'm not inclined to backport a stabilization.

**Comment by nikomatsakis** [comment]
Note: doc pr is here: https://github.com/rust-lang-nursery/reference/pull/12

**Comment by petrochenkov** [comment]
Right now this PR turns all private-in-public errors for `pub(restricted)` into warnings (since they are not in the "old set").
I'd prefer @cramertj to implement https://github.com/rust-lang/rust/pull/40556#discussion_r106292150 before merging.

@bors r-

**Comment by nikomatsakis** [comment]
Hmm. I see. OK, that probably makes sense.

**Comment by ubsan** [comment]
I disagree with this stabilization.

**Comment by nikomatsakis** [comment]
Removing beta-nominated. Seems like there is no need to backport a stabilization, no great urgency.

**Comment by ubsan** [comment]
Okay, I remove my disagreement, actually (sorry). I read more of the reasoning behind the need for `in path`, and I can agree that it is useful. (sorry sorry sorry)

**Comment by petrochenkov** [comment]
@cramertj 
A simpler and more precise solution - just walk all the crate's HIR and check if `pub(restricted)` is ever used.
This can be done in advance even if no errors happen, (almost) no-op HIR walk is super fast and won't affect performance (and this is all temporary anyway).

**Comment by cramertj** [comment]
@petrochenkov I was just finishing up piping `&'tcx hir::Visibility` through the whole thing and shaking my head at how much churn wound up being necessary. This sounds like a much better idea.

Just to make sure I understood you: the process is to add a new visitor which runs in `librustc_privacy/lib.rs::check_crate` and returns a bool of whether or not `pub_restricted` was used. If it was, all "private in public" warnings become errors?

**Comment by petrochenkov** [comment]
>Just to make sure I understood you: the process is to add a new visitor which runs in librustc_privacy/lib.rs::check_crate and returns a bool of whether or not pub_restricted was used. If it was, all "private in public" warnings become errors?

Exactly.
That would be equivalent to the previous behavior (modulo unused `#![feature(pub_restricted)]`s).

**Pull request merged**